commons/src/main/java/org/infinispan/commons/util/ProcessorInfo.java

```diff
@@ -0,0 +1,144 @@
+package org.infinispan.commons.util;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.Locale;
+import java.util.stream.Stream;
+
+/**
+ * Provides general information about the processors on this host.
+ *
+ * @author Jason T. Greene
+ * @author Tristan Tarrant
+ */
+public class ProcessorInfo {
+   private static final byte[] BITS = new byte[]{0, 1, 1, 2, 1, 2, 2, 3, 1, 2, 2, 3, 2, 3, 3, 4};
+   private static final String CPUS_ALLOWED = "Cpus_allowed:";
+   private static final String CGROUP_STATUS_FILE = "/proc/cgroups";
+   private static final String CPU_CFS_PERIOD_US = "/sys/fs/cgroup/cpu/cpu.cfs_period_us";
+   private static final String CPU_CFS_QUOTA_US = "/sys/fs/cgroup/cpu/cpu.cfs_quota_us";
+   private static final String CPU_SHARES = "/sys/fs/cgroup/cpu/cpu.shares";
+
+   private ProcessorInfo() {
+   }
+
+   /**
+    * Returns the number of processors available to this process. On most operating systems this method
+    * simply delegates to {@link Runtime#availableProcessors()}. However, on Linux, this strategy
+    * is insufficient, since the JVM does not take into consideration the process' CPU set affinity and the CGroups
+    * quota/period assignment. Therefore this method will analyze the Linux proc filesystem
+    * to make the determination. Since the CPU affinity of a process can be change at any time, this method does
+    * not cache the result. Calls should be limited accordingly.
+    * <br>
+    * The number of available processors can be overridden via the system property <tt>infinispan.activeprocessorcount</tt>,
+    * e.g. <tt>java -Dinfinispan.activeprocessorcount=4 ...</tt>. Note that this value cannot exceed the actual number
+    * of available processors.
+    * <br>
+    * Since Java 10, this can also be achieved via the VM flag <tt>-XX:ActiveProcessorCount=xx</tt>.
+    * <br>
+    * Note that on Linux, both SMT units (Hyper-Threading) and CPU cores are counted as a processor.
+    *
+    * @return the available processors on this system.
+    */
+   public static int availableProcessors() {
+      if (System.getSecurityManager() != null) {
+         return AccessController.doPrivileged((PrivilegedAction<Integer>) () -> Integer.valueOf(determineProcessors())).intValue();
+      }
+
+      return determineProcessors();
+   }
+
+   private static int determineProcessors() {
+      int javaProcs = Runtime.getRuntime().availableProcessors();
+      int userProcs = Integer.getInteger("infinispan.activeprocessorcount", javaProcs);
```

Comment:
Shouldn't we always use the value returned from the setting "infinispan.activeprocessorcount" ?

Reply:
it does so only if it is less than the available processors.

Reply:
Okay, from the name I assumed it would be use this setting always type of thing. Just side note but we non linux would always use this value if it is set btw.